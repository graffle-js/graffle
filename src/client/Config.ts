import type { ExecutionResult } from 'graphql'
import type { GraphQLExecutionResultError } from '../lib/graphql.js'
import type { SetProperty } from '../lib/prelude.js'

export type ReturnModeType =
  | ReturnModeTypeGraphQL
  | ReturnModeTypeData
  | ReturnModeTypeDataAndSchemaErrors
  | ReturnModeTypeDataAllErrors

export type ReturnModeTypeBase = ReturnModeTypeGraphQL | ReturnModeTypeData | ReturnModeTypeDataAllErrors

export type ReturnModeTypeGraphQL = 'graphql'

export type ReturnModeTypeData = 'data'

export type ReturnModeTypeDataAllErrors = 'dataAndAllErrors'

export type ReturnModeTypeDataAndSchemaErrors = 'dataAndSchemaErrors'

export type OptionsInput = {
  returnMode: ReturnModeType | undefined
}

export type OptionsInputDefaults = {
  returnMode: 'data'
}

export type Config = {
  returnMode: ReturnModeType
}

export type ApplyInputDefaults<Input extends OptionsInput> = {
  [Key in keyof OptionsInputDefaults]: undefined extends Input[Key] ? OptionsInputDefaults[Key]
    : Exclude<Input[Key], undefined>
}

// dprint-ignore
export type ReturnMode<$Config extends Config, $Data, $DataRaw = undefined> =
  $Config['returnMode'] extends 'graphql'     ? ExecutionResult<$DataRaw extends undefined ? $Data : $DataRaw> :
  $Config['returnMode'] extends 'data'        ? $Data :
                                                $Data | GraphQLExecutionResultError

export type OrThrowifyConfig<$Config extends Config> = $Config['returnMode'] extends 'graphql' ? $Config
  : SetProperty<$Config, 'returnMode', 'data'>
