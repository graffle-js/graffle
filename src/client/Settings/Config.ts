import type { GraphQLSchema } from 'graphql'

export type OutputChannel = 'throw' | 'return'

export type OutputChannelConfig = 'throw' | 'return' | 'default'

export type ErrorCategory = 'execution' | 'other'

export const readConfigErrorCategoryOutputChannel = (
  config: Config,
  errorCategory: ErrorCategory,
): OutputChannel | false => {
  if (config.output.errors[errorCategory] === `default`) {
    return config.output.defaults.errorChannel
  }
  return config.output.errors[errorCategory]
}

export const traditionalGraphqlOutput = {
  defaults: { errorChannel: `throw` },
  envelope: { enabled: true, errors: { execution: true, other: false } },
  errors: { execution: `default`, other: `default` },
} satisfies OutputConfig

export const traditionalGraphqlOutputThrowing: OutputConfig = {
  ...traditionalGraphqlOutput,
  envelope: {
    ...traditionalGraphqlOutput.envelope,
    errors: {
      ...traditionalGraphqlOutput.envelope.errors,
      execution: false,
    },
  },
}

export const isContextConfigTraditionalGraphQLOutput = (config: Config) => {
  return config.output.envelope.enabled && config.output.envelope.errors.execution
    && !config.output.envelope.errors.other
}

export type OutputConfig = {
  defaults: {
    errorChannel: OutputChannel
  }
  envelope: {
    enabled: boolean
    errors: {
      execution: boolean
      other: boolean
    }
  }
  errors: {
    execution: OutputChannelConfig
    other: OutputChannelConfig
  }
}

export const outputConfigDefault: OutputConfigDefault = {
  defaults: {
    errorChannel: `throw`,
  },
  envelope: {
    enabled: false,
    errors: {
      execution: true,
      other: false,
    },
  },
  errors: {
    execution: `default`,
    other: `default`,
  },
}

export type OutputConfigDefault = {
  defaults: {
    errorChannel: 'throw'
  }
  envelope: {
    enabled: false
    errors: {
      execution: true
      other: false
    }
  }
  errors: {
    execution: 'default'
    other: 'default'
  }
}

export interface TransportConfigMemory {
  schema: GraphQLSchema
}

export type Config = {
  output: OutputConfig
}
