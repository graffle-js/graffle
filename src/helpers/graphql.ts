/**
 * Clean a GraphQL document to send it via a GET query
 */
export const cleanQuery = (str: string): string => str.replace(/([\s,]|#[^\n\r]+)+/g, ` `).trim()
