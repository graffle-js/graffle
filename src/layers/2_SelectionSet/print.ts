import { RootTypeName } from '../../lib/graphql.js'
import { assertArray, assertObject, entriesStrict, lowerCaseFirstLetter } from '../../lib/prelude.js'
import { Schema } from '../1_Schema/__.js'
import { readMaybeThunk } from '../1_Schema/core/helpers.js'
import type { Config } from '../6_client/Settings/Config.js'
import type { Indicator } from './_indicator.js'
import { Directive } from './Directive/__.js'
import { SpecialKeyPrefixes } from './helpers.js'
import { isIndicator, isPositiveIndicator } from './indicator.js'
import { parseClientOn, toGraphQLOn } from './on.js'
import { parseClientFieldItem } from './runtime/FieldItem.js'
import { createFieldName } from './runtime/FieldName.js'
import { type ArgValue, normalizeAlias, type ObjectLike, type SS } from './types.js'

type FieldValue = SS | Indicator.Indicator

export interface Context {
  schemaIndex: Schema.Index
  config: Config
}

export const rootType = (
  context: Context,
  rootObjectDef: Schema.Output.RootType,
  selectionSet: ObjectLike,
  operationName: string = ``,
) => {
  const operationTypeName = lowerCaseFirstLetter(rootObjectDef.fields.__typename.type.type)
  return `${operationTypeName} ${operationName} { ${printObjectLikeField(context, rootObjectDef, selectionSet)} }`
}

const argValue = (
  context: Context,
  schemaArgTypeMaybeThunk: Schema.Input.Any,
  value: ArgValue,
): string => {
  if (value === null) return String(null) // todo could check if index agrees is nullable.

  const schemaArgType = readMaybeThunk(schemaArgTypeMaybeThunk)

  switch (schemaArgType.kind) {
    case `nullable`:
      return argValue(context, schemaArgType.type, value)
    case `list`: {
      assertArray(value)
      const values = value.map(_ => argValue(context, schemaArgType.type, _ as ArgValue))
      return `[${values.join(`, `)}]`
    }
    case `InputObject`: {
      assertObject(value)
      const entries = Object.entries(value).map(([argName_, argValue_]) => {
        const schemaArgField = schemaArgType.fields[argName_] as Schema.Input.Field | undefined
        if (!schemaArgField) throw new Error(`Arg not found: ${argName_}`)
        return [argName_, argValue(context, schemaArgField.type, argValue_)]
      })
      return `{ ${entries.map(([k, v]) => `${k!}: ${v!}`).join(`, `)} }`
    }
    case `Enum`: {
      return String(value)
    }
    case `Scalar`: {
      // @ts-expect-error fixme
      return JSON.stringify(schemaArgType.codec.encode(value))
    }
    default:
      throw new Error(`Unsupported arg kind: ${JSON.stringify(schemaArgType)}`)
  }
}

const pruneNonSelections = (ss: SS) => {
  const entries = Object.entries(ss)
  const selectEntries = entries.filter(_ => !_[0].startsWith(`$`))
  return Object.fromEntries(selectEntries)
}

export const printObjectLikeField = (
  context: Context,
  schemaItem: Schema.Object$2 | Schema.Union | Schema.Interface,
  fieldValue: FieldValue,
): string => {
  // todo fixme
  const expressions = separateExpressions(fieldValue as SS)

  const printedFieldGroups = expressions.fieldGroups.map(group => {
    return printFragmentInlineFieldGroup(context, schemaItem, group)
  }).join(`\n`) + `\n`

  switch (schemaItem.kind) {
    case `Object`: {
      const rootTypeName = (RootTypeName as Record<string, RootTypeName>)[schemaItem.fields.__typename.type.type]
        ?? null
      const printedSchemaFields = expressions.schemaFields.map(([clientFieldName, ss]) => {
        const fieldName = createFieldName(clientFieldName)
        const schemaField = schemaItem.fields[fieldName.value]
        if (!schemaField) throw new Error(`Field ${clientFieldName} not found in schema object`)
        /**
         * Inject __typename field for result fields that are missing it.
         */
        // dprint-ignore
        if (rootTypeName && context.config.output.errors.schema !== false && context.schemaIndex.error.rootResultFields[rootTypeName][fieldName.value]) {
          (ss as Record<string, boolean>)[`__typename`] = true
        }
        return printField(context, schemaField, { name: fieldName.value, value: ss })
      }).join(`\n`) + `\n`

      return printedSchemaFields + printedFieldGroups
    }
    case `Interface`: {
      const printedSchemaFields = expressions.schemaFields.map(([ClientFieldName, ss]) => {
        const fieldItem = parseClientFieldItem(ClientFieldName)

        switch (fieldItem._tag) {
          case `FieldName`: {
            if (fieldItem.value === `__typename`) {
              return `${fieldItem.value} ${printDirectives(ss)}`
            }
            const schemaField = schemaItem.fields[fieldItem.value]
            if (!schemaField) throw new Error(`Field ${ClientFieldName} not found in schema object`)
            return printField(context, schemaField, { name: fieldItem.value, value: ss })
          }
          case `On`: {
            const schemaObject = context.schemaIndex[`objects`][fieldItem.typeOrFragmentName]
            if (!schemaObject) throw new Error(`Fragment ${fieldItem.typeOrFragmentName} not found in schema`)
            return `${toGraphQLOn(fieldItem)} ${printDirectives(ss)} { ${
              printObjectLikeField(context, schemaObject, ss)
            } }`
          }
          default: {
            throw new Error(`Unknown field item tag`)
          }
        }
      }).join(`\n`) + `\n`

      return printedSchemaFields
    }
    case `Union`: {
      const printedSchemaFields = expressions.schemaFields.map(([fieldExpression, ss]) => {
        const fieldItem = parseClientFieldItem(fieldExpression)
        switch (fieldItem._tag) {
          case `FieldName`: {
            if (fieldItem.value === `__typename`) {
              return `${fieldItem.value} ${printDirectives(ss)}`
            }
            console.log(fieldItem)
            // todo
            throw new Error(`todo resolve common interface fields from unions`)
          }
          case `On`: {
            const schemaObject = context.schemaIndex[`objects`][fieldItem.typeOrFragmentName]
            if (!schemaObject) throw new Error(`Fragment ${fieldItem.typeOrFragmentName} not found in schema`)
            // if (isIndicator(ss)) throw new Error(`Union field must have selection set`)
            return `${toGraphQLOn(fieldItem)} ${printDirectives(ss)} { ${
              // @ts-expect-error fixme
              printObjectLikeField(context, schemaObject, pruneNonSelections(ss))} }`
          }
          default: {
            throw new Error(`Unknown field item tag`)
          }
        }
      }).join(`\n`) + `\n`

      return printedSchemaFields
    }
    default:
      throw new Error(`Unknown schema item kind`)
  }
}

export const printFragmentInlineFieldGroup = (
  context: Context,
  type: Schema.Object$2 | Schema.Interface | Schema.Union,
  group: SS,
) => {
  const directivesPrinted = printDirectives(group)
  return `... ${directivesPrinted} {\n${printObjectLikeField(context, type, group)}\n}`
}

export const printField = (
  context: Context,
  schemaField: Schema.SomeField,
  field: { name: string; value: FieldValue },
): string => {
  const aliases = normalizeAlias(field.value)

  if (aliases) {
    return aliases.map(alias => {
      return `${alias[0]}: ${field.name} ${printFieldValue(context, schemaField, alias[1] as FieldValue)}`
    }).join(`\n`)
  }

  return `${field.name} ${printFieldValue(context, schemaField, field.value)}`
}

const printFieldValue = (
  context: Context,
  schemaField: Schema.SomeField,
  fieldValue: null | FieldValue,
): string => {
  if (fieldValue === null) return `null` // todo test this case

  if (isIndicator(fieldValue)) return ``
  const keys = separateExpressions(fieldValue)

  const directives = printDirectives(fieldValue)
  const args = printArgs(context, schemaField, fieldValue)

  if (keys.schemaFields.length === 0 && keys.fieldGroups.length === 0) {
    return `${args} ${directives}`
  }

  // eslint-disable-next-line
  // @ts-ignore ID error
  const schemaNamedOutputType = Schema.Output.unwrapToNamed(schemaField.type) as Schema.Object$2
  const schemaNamedOutputType_ = readMaybeThunk(schemaNamedOutputType)

  const fieldGroups = keys.fieldGroups.map(group => {
    return printFragmentInlineFieldGroup(context, schemaNamedOutputType_, group)
  }).join(`\n`) + `\n`

  const selection = Object.fromEntries(keys.schemaFields) as ObjectLike

  return `${args} ${directives} {
		${fieldGroups}
		${printObjectLikeField(context, schemaNamedOutputType_, selection)}
	}`
}

const printArgs = (context: Context, schemaField: Schema.SomeField, ss: Indicator.Indicator | SS) => {
  if (isIndicator(ss)) return ``

  const { $ } = ss
  if ($ === undefined) return ``

  const schemaArgs = schemaField.args
  if (!schemaArgs) throw new Error(`Field has no args`)

  const argEntries = Object.entries($)
  if (argEntries.length === 0) return ``

  return `(${
    argEntries.map(([argFieldName, v]) => {
      const schemaArgField = schemaArgs.fields[argFieldName] as Schema.Input.Field | undefined
      if (!schemaArgField) throw new Error(`Arg field ${argFieldName} not found in schema.`)
      const valueEncoded = argValue(context, schemaArgField.type, v)
      return `${argFieldName}: ${valueEncoded}`
    }).join(`, `)
  })`
}

export const resolveFragmentInlineOn = (field: string): string => {
  const on = parseClientOn(field)
  if (on) return toGraphQLOn(on)
  return field
}

const printDirectives = (fieldValue: FieldValue): string => {
  if (isIndicator(fieldValue)) return ``

  return entriesStrict({
    $skip: fieldValue.$skip,
    $include: fieldValue.$include,
  }).reduce((code, [field, input]) => {
    const def = Directive.fieldToDef[field]
    code += Directive.Print.toGraphQLDirective(def.create(input))
    return code
  }, ``)
}

const separateExpressions = (ss: SS) => {
  const { ___, $, ...ssRest } = ss

  const fieldGroupOrGroups = ___ as SS | SS[] | undefined
  const fieldGroups = fieldGroupOrGroups
    ? Array.isArray(fieldGroupOrGroups) ? fieldGroupOrGroups : [fieldGroupOrGroups]
    : []

  const restApplicable = Object.entries(ssRest)
    .filter(([_, ss]) => isPositiveIndicator(ss)) as [
      string,
      FieldValue,
    ][]

  const directives = restApplicable.filter(([k, _]) => k.startsWith(SpecialKeyPrefixes.argumentsOrDirective))

  const schemaFields = restApplicable.filter(([k, _]) => !k.startsWith(SpecialKeyPrefixes.argumentsOrDirective))

  return {
    fieldGroups,
    args: $,
    directives,
    schemaFields,
  }
}
