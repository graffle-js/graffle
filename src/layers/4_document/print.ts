import { operationTypeNameToRootTypeName, OperationTypes } from '../../lib/graphql.js'
import { SelectionSet } from '../2_SelectionSet/__.js'
import type { Context, DocumentObject } from '../2_SelectionSet/print.js'

// todo this is currently unused by graffle internally. Remove?
export const print = (
  context: Context,
  document: DocumentObject,
) => {
  const operations = [
    ...(Object.entries(document.query || {}).map(([operationName, selectionSet]) => ({
      operationName,
      selectionSet,
      operationType: OperationTypes.query,
    }))),
    ...(Object.entries(document.mutation || {}).map(([operationName, selectionSet]) => ({
      operationName,
      selectionSet,
      operationType: OperationTypes.mutation,
    }))),
  ]

  return operations.map(({ operationName, selectionSet, operationType }) => {
    const rootType = operationTypeNameToRootTypeName[operationType]
    const rootTypeDocument = selectionSet

    const schemaRootType = context.schemaIndex[`Root`][rootType]
    if (!schemaRootType) throw new Error(`Schema has no ${rootType} root type`)

    const documentString = SelectionSet.Print.resolveRootType(
      context,
      schemaRootType,
      rootTypeDocument,
      operationName,
    )

    return documentString
  }).join(`\n\n`)
}
