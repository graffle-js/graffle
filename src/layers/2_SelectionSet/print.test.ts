import { parse, print } from 'graphql'
import { describe, expect, test } from 'vitest'
import { db } from '../../../tests/_/schemas/db.js'
import { $Index as schemaIndex } from '../../../tests/_/schemas/kitchen-sink/graffle/modules/SchemaRuntime.js'
import type * as SelectionSets from '../../../tests/_/schemas/kitchen-sink/graffle/modules/SelectionSets.js'
import { outputConfigDefault } from '../6_client/Settings/Config.js'
import type { Context } from './print.js'
import { rootType } from './print.js'

// eslint-disable-next-line
// @ts-ignore
const s = (selectionSet: SelectionSets.Query) => selectionSet

const testEachArgs = [
  `Query`,
  (
    ...args: [SelectionSets.Query] | [
      description: string,
      ss: SelectionSets.Query,
    ]
  ) => {
    const [description, ss] = args.length === 1 ? [undefined, args[0]] : args
    const context: Context = {
      schemaIndex,
      config: {
        output: outputConfigDefault,
        transport: { type: `memory`, config: { methodMode: `post` } },
        name: schemaIndex[`name`],

        initialInput: {} as any,
      },
    }
    const graphqlDocumentString = rootType(context, schemaIndex[`Root`][`Query`], ss as any)
    // console.log(graphqlDocumentString)
    // Should parse, ensures is syntactically valid graphql document.
    const document = parse(graphqlDocumentString)
    const graphqlDocumentStringFormatted = print(document)
    const beforeAfter = `\n`
      + JSON.stringify(ss, null, 2)
      + `\n--------------\n`
      + graphqlDocumentStringFormatted
      + `\n`
    expect(beforeAfter).toMatchSnapshot(description)
  },
] as const

describe(`inline fragment field group`, () => {
  test.each([
    [s({ ___: { __typename: true } })],
    [s({ ___: [{ __typename: true }, { abcEnum: true }] })],
    [s({ interface: { ___: { __typename: true } } })],
    [s({ unionFooBar: { ___: { __typename: true } } })],
    [s({ ___: { __typename: true, $include: true } })],
  ])(...testEachArgs)
})

describe(`enum`, () => {
  test.each([
    [s({ result: { $: { case: `Object1` }, __typename: true } })],
  ])(...testEachArgs)
})

describe(`union`, () => {
  test.each([
    [s({ unionFooBar: { __typename: true } })],
    [s({ unionFooBar: { ___on_Bar: { int: true } } })],
    [s({ unionFooBar: { ___on_Bar: { $skip: true, int: true } } })],
    // s({ unionFooBar: { __on_Bar: {} } }), // todo should be static type error
  ])(...testEachArgs)
})

describe(`alias`, () => {
  test.each([
    [s({ id: [`x`, true] })],
    [s({ id: [[`x`, true], [`id2`, true]] })],
    [s({ id: [`x`, { $skip: true }] })],
    [s({ object: [`x`, { $skip: true, id: true }] })],
  ])(...testEachArgs)
})

describe(`args`, () => {
  test.each([
    [s({ stringWithArgs: { $: { boolean: true, float: 1 } } })],
    [s({ stringWithArgs: { $: {} } })],
    // s({ objectWithArgs: { $: { id: `` } } }), // todo should be static error
    // s({ objectWithArgs: { $: {} } }), // todo should be static error
    [s({ objectWithArgs: { $: { id: `` }, id: true } })],
    [s({ objectWithArgs: { $: {}, id: true } })],
  ])(...testEachArgs)
})

describe(`$include`, () => {
  test.each([
    [s({ object: { $include: true, id: true } })],
    [s({ object: { $include: false, id: true } })],
    [s({ object: { $include: undefined, id: true } })],
    [s({ object: { $include: { if: true }, id: true } })],
    [s({ object: { $include: { if: false }, id: true } })],
    [s({ object: { $include: { if: undefined }, id: true } })],
    [s({ object: { $include: {}, id: true } })],
  ])(...testEachArgs)
})

describe(`$skip`, () => {
  test.each([
    [s({ object: { $skip: true, id: true } })],
    [s({ object: { $skip: false, id: true } })],
    [s({ object: { $skip: undefined, id: true } })],
    [s({ object: { $skip: { if: true }, id: true } })],
    [s({ object: { $skip: { if: false }, id: true } })],
    [s({ object: { $skip: { if: undefined }, id: true } })],
    [s({ object: { $skip: {}, id: true } })],
  ])(...testEachArgs)
})

// describe(`$defer`, () => {
//   test.each([
//     [s({ object: { $defer: true, id: true } })],
//     [s({ object: { $defer: false, id: true } })],
//     [s({ object: { $defer: undefined, id: true } })],
//     [s({ object: { $defer: { if: true }, id: true } })],
//     [s({ object: { $defer: { if: false }, id: true } })],
//     [s({ object: { $defer: { if: undefined }, id: true } })],
//     [s({ object: { $defer: {}, id: true } })],
//     [s({ object: { $defer: { label: `foobar` }, id: true } })],
//   ])(...testEachArgs)
// })

// describe(`$stream`, () => {
//   test.each([
//     [s({ object: { $stream: true, id: true } })],
//     [s({ object: { $stream: false, id: true } })],
//     [s({ object: { $stream: undefined, id: true } })],
//     [s({ object: { $stream: { if: true }, id: true } })],
//     [s({ object: { $stream: { if: false }, id: true } })],
//     [s({ object: { $stream: { if: undefined }, id: true } })],
//     [s({ object: { $stream: {}, id: true } })],
//     [s({ object: { $stream: { label: `foobar` }, id: true } })],
//     [s({ object: { $stream: { initialCount: 5 }, id: true } })],
//   ])(...testEachArgs)
// })

describe(`other`, () => {
  test.each([
    [s({ __typename: true })],
    [s({ string: true })],
    // [s({ string: 1 })],
    // s({ string: false }), // todo should be static error
    [s({ id: true, string: false })],
    // [s({ id: true, string: 0 })],
    [s({ id: true, string: undefined })],
    [s({ object: { id: true } })],
    [s({ objectNested: { object: { string: true, id: true, int: false } } })],
    [s({ objectNested: { object: { string: true, id: true, int: { $skip: true } } } })],
  ])(...testEachArgs)
})

describe(`args`, () => {
  // dprint-ignore
  describe(`custom scalars`, () => {
    test.each([
      [`arg field`, s({ dateArg: { $: { date: db.date0 } } })],
      [`arg field in non-null`,s({ dateArgNonNull: { $: { date: db.date0 } } })], 
      [`arg field in list`,s({ dateArgList: { $: { date: [db.date0, new Date(1)] } } })], 
      [`arg field in list (null)`,s({ dateArgList: { $: { date: null } } })],
      [`arg field in non-null list (with list)`,s({ dateArgNonNullList: { $: { date: [db.date0, new Date(1)] } } })],
      [`arg field in non-null list (with null)`,s({ dateArgNonNullList: { $: { date: [null, db.date0] } } })],
      [`arg field in non-null list non-null`,s({ dateArgNonNullListNonNull: { $: { date: [db.date0, new Date(1)] } } })],
      [`input object field`,s({ dateArgInputObject: { $: { input: { idRequired: ``, dateRequired: db.date0, date: new Date(1) } } } })],
      [`nested input object field`, s({ InputObjectNested: { $: { input: { InputObject: { idRequired: ``, dateRequired: db.date0, date: new Date(1) } } } } })]
    ] as const)(...testEachArgs)
  })
})
