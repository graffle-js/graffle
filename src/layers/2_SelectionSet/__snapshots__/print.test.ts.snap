// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`$include > Query 1`] = `
"
{
  "object": {
    "$include": true,
    "id": true
  }
}
--------------
{
  object @include(if: true) {
    id
  }
}
"
`;

exports[`$include > Query 2`] = `
"
{
  "object": {
    "$include": false,
    "id": true
  }
}
--------------
{
  object @include(if: false) {
    id
  }
}
"
`;

exports[`$include > Query 3`] = `
"
{
  "object": {
    "id": true
  }
}
--------------
{
  object {
    id
  }
}
"
`;

exports[`$include > Query 4`] = `
"
{
  "object": {
    "$include": {
      "if": true
    },
    "id": true
  }
}
--------------
{
  object @include(if: true) {
    id
  }
}
"
`;

exports[`$include > Query 5`] = `
"
{
  "object": {
    "$include": {
      "if": false
    },
    "id": true
  }
}
--------------
{
  object @include(if: false) {
    id
  }
}
"
`;

exports[`$include > Query 6`] = `
"
{
  "object": {
    "$include": {},
    "id": true
  }
}
--------------
{
  object @include(if: true) {
    id
  }
}
"
`;

exports[`$include > Query 7`] = `
"
{
  "object": {
    "$include": {},
    "id": true
  }
}
--------------
{
  object @include(if: true) {
    id
  }
}
"
`;

exports[`$skip > Query 1`] = `
"
{
  "object": {
    "$skip": true,
    "id": true
  }
}
--------------
{
  object @skip(if: true) {
    id
  }
}
"
`;

exports[`$skip > Query 2`] = `
"
{
  "object": {
    "$skip": false,
    "id": true
  }
}
--------------
{
  object @skip(if: false) {
    id
  }
}
"
`;

exports[`$skip > Query 3`] = `
"
{
  "object": {
    "id": true
  }
}
--------------
{
  object {
    id
  }
}
"
`;

exports[`$skip > Query 4`] = `
"
{
  "object": {
    "$skip": {
      "if": true
    },
    "id": true
  }
}
--------------
{
  object @skip(if: true) {
    id
  }
}
"
`;

exports[`$skip > Query 5`] = `
"
{
  "object": {
    "$skip": {
      "if": false
    },
    "id": true
  }
}
--------------
{
  object @skip(if: false) {
    id
  }
}
"
`;

exports[`$skip > Query 6`] = `
"
{
  "object": {
    "$skip": {},
    "id": true
  }
}
--------------
{
  object @skip(if: true) {
    id
  }
}
"
`;

exports[`$skip > Query 7`] = `
"
{
  "object": {
    "$skip": {},
    "id": true
  }
}
--------------
{
  object @skip(if: true) {
    id
  }
}
"
`;

exports[`alias > Query 1`] = `
"
{
  "id": [
    "x",
    true
  ]
}
--------------
{
  x: id
}
"
`;

exports[`alias > Query 2`] = `
"
{
  "id": [
    [
      "x",
      true
    ],
    [
      "id2",
      true
    ]
  ]
}
--------------
{
  x: id
  id2: id
}
"
`;

exports[`alias > Query 3`] = `
"
{
  "id": [
    "x",
    {
      "$skip": true
    }
  ]
}
--------------
{
  x: id @skip(if: true)
}
"
`;

exports[`alias > Query 4`] = `
"
{
  "object": [
    "x",
    {
      "$skip": true,
      "id": true
    }
  ]
}
--------------
{
  x: object @skip(if: true) {
    id
  }
}
"
`;

exports[`args > Query 1`] = `
"
{
  "stringWithArgs": {
    "$": {
      "boolean": true,
      "float": 1
    }
  }
}
--------------
{
  stringWithArgs(boolean: true, float: 1)
}
"
`;

exports[`args > Query 2`] = `
"
{
  "stringWithArgs": {
    "$": {}
  }
}
--------------
{
  stringWithArgs
}
"
`;

exports[`args > Query 3`] = `
"
{
  "objectWithArgs": {
    "$": {
      "id": ""
    },
    "id": true
  }
}
--------------
{
  objectWithArgs(id: "") {
    id
  }
}
"
`;

exports[`args > Query 4`] = `
"
{
  "objectWithArgs": {
    "$": {},
    "id": true
  }
}
--------------
{
  objectWithArgs {
    id
  }
}
"
`;

exports[`args > custom scalars > Query > arg field 1`] = `
"
{
  "dateArg": {
    "$": {
      "date": "1970-01-01T00:00:00.000Z"
    }
  }
}
--------------
{
  dateArg(date: "1970-01-01T00:00:00.000Z")
}
"
`;

exports[`args > custom scalars > Query > arg field in list (null) 1`] = `
"
{
  "dateArgList": {
    "$": {
      "date": null
    }
  }
}
--------------
{
  dateArgList(date: null)
}
"
`;

exports[`args > custom scalars > Query > arg field in list 1`] = `
"
{
  "dateArgList": {
    "$": {
      "date": [
        "1970-01-01T00:00:00.000Z",
        "1970-01-01T00:00:00.001Z"
      ]
    }
  }
}
--------------
{
  dateArgList(date: ["1970-01-01T00:00:00.000Z", "1970-01-01T00:00:00.001Z"])
}
"
`;

exports[`args > custom scalars > Query > arg field in non-null 1`] = `
"
{
  "dateArgNonNull": {
    "$": {
      "date": "1970-01-01T00:00:00.000Z"
    }
  }
}
--------------
{
  dateArgNonNull(date: "1970-01-01T00:00:00.000Z")
}
"
`;

exports[`args > custom scalars > Query > arg field in non-null list (with list) 1`] = `
"
{
  "dateArgNonNullList": {
    "$": {
      "date": [
        "1970-01-01T00:00:00.000Z",
        "1970-01-01T00:00:00.001Z"
      ]
    }
  }
}
--------------
{
  dateArgNonNullList(
    date: ["1970-01-01T00:00:00.000Z", "1970-01-01T00:00:00.001Z"]
  )
}
"
`;

exports[`args > custom scalars > Query > arg field in non-null list (with null) 1`] = `
"
{
  "dateArgNonNullList": {
    "$": {
      "date": [
        null,
        "1970-01-01T00:00:00.000Z"
      ]
    }
  }
}
--------------
{
  dateArgNonNullList(date: [null, "1970-01-01T00:00:00.000Z"])
}
"
`;

exports[`args > custom scalars > Query > arg field in non-null list non-null 1`] = `
"
{
  "dateArgNonNullListNonNull": {
    "$": {
      "date": [
        "1970-01-01T00:00:00.000Z",
        "1970-01-01T00:00:00.001Z"
      ]
    }
  }
}
--------------
{
  dateArgNonNullListNonNull(
    date: ["1970-01-01T00:00:00.000Z", "1970-01-01T00:00:00.001Z"]
  )
}
"
`;

exports[`args > custom scalars > Query > input object field 1`] = `
"
{
  "dateArgInputObject": {
    "$": {
      "input": {
        "idRequired": "",
        "dateRequired": "1970-01-01T00:00:00.000Z",
        "date": "1970-01-01T00:00:00.001Z"
      }
    }
  }
}
--------------
{
  dateArgInputObject(
    input: {idRequired: "", dateRequired: "1970-01-01T00:00:00.000Z", date: "1970-01-01T00:00:00.001Z"}
  )
}
"
`;

exports[`args > custom scalars > Query > nested input object field 1`] = `
"
{
  "InputObjectNested": {
    "$": {
      "input": {
        "InputObject": {
          "idRequired": "",
          "dateRequired": "1970-01-01T00:00:00.000Z",
          "date": "1970-01-01T00:00:00.001Z"
        }
      }
    }
  }
}
--------------
{
  InputObjectNested(
    input: {InputObject: {idRequired: "", dateRequired: "1970-01-01T00:00:00.000Z", date: "1970-01-01T00:00:00.001Z"}}
  )
}
"
`;

exports[`enum > Query 1`] = `
"
{
  "result": {
    "$": {
      "case": "Object1"
    },
    "__typename": true
  }
}
--------------
{
  result(case: Object1) {
    __typename
  }
}
"
`;

exports[`inline fragment field group > Query 1`] = `
"
{
  "___": {
    "__typename": true
  }
}
--------------
{
  ... {
    __typename
  }
}
"
`;

exports[`inline fragment field group > Query 2`] = `
"
{
  "___": [
    {
      "__typename": true
    },
    {
      "abcEnum": true
    }
  ]
}
--------------
{
  ... {
    __typename
  }
  ... {
    abcEnum
  }
}
"
`;

exports[`inline fragment field group > Query 3`] = `
"
{
  "interface": {
    "___": {
      "__typename": true
    }
  }
}
--------------
{
  interface {
    ... {
      __typename
    }
  }
}
"
`;

exports[`inline fragment field group > Query 4`] = `
"
{
  "unionFooBar": {
    "___": {
      "__typename": true
    }
  }
}
--------------
{
  unionFooBar {
    ... {
      __typename
    }
  }
}
"
`;

exports[`inline fragment field group > Query 5`] = `
"
{
  "___": {
    "__typename": true,
    "$include": true
  }
}
--------------
{
  ... @include(if: true) {
    __typename
  }
}
"
`;

exports[`other > Query 1`] = `
"
{
  "__typename": true
}
--------------
{
  __typename
}
"
`;

exports[`other > Query 2`] = `
"
{
  "string": true
}
--------------
{
  string
}
"
`;

exports[`other > Query 3`] = `
"
{
  "id": true,
  "string": false
}
--------------
{
  id
}
"
`;

exports[`other > Query 4`] = `
"
{
  "id": true
}
--------------
{
  id
}
"
`;

exports[`other > Query 5`] = `
"
{
  "object": {
    "id": true
  }
}
--------------
{
  object {
    id
  }
}
"
`;

exports[`other > Query 6`] = `
"
{
  "objectNested": {
    "object": {
      "string": true,
      "id": true,
      "int": false
    }
  }
}
--------------
{
  objectNested {
    object {
      string
      id
    }
  }
}
"
`;

exports[`other > Query 7`] = `
"
{
  "objectNested": {
    "object": {
      "string": true,
      "id": true,
      "int": {
        "$skip": true
      }
    }
  }
}
--------------
{
  objectNested {
    object {
      string
      id
      int @skip(if: true)
    }
  }
}
"
`;

exports[`union > Query 1`] = `
"
{
  "unionFooBar": {
    "__typename": true
  }
}
--------------
{
  unionFooBar {
    __typename
  }
}
"
`;

exports[`union > Query 2`] = `
"
{
  "unionFooBar": {
    "___on_Bar": {
      "int": true
    }
  }
}
--------------
{
  unionFooBar {
    ... on Bar {
      int
    }
  }
}
"
`;

exports[`union > Query 3`] = `
"
{
  "unionFooBar": {
    "___on_Bar": {
      "$skip": true,
      "int": true
    }
  }
}
--------------
{
  unionFooBar {
    ... on Bar @skip(if: true) {
      int
    }
  }
}
"
`;
